File: SEM2/ALGO/mergesort.cpp
#include<iostream>
#include<vector>
using namespace std;
void displayArray(vector<int> &array){
    cout<<"Displaying Your Array..."<<endl;
    for (int i = 0; i < array.size(); i++){
        cout<<array[i]<<" ";
    }
    cout<<endl;
}

vector<int> merge(vector<int> &array1,vector<int> &array2){
    vector<int> res(array1.size()+array2.size());
    int p1 = 0;
    int p2 = 0;
    int index = 0;
    while (p1<array1.size() && p2<array2.size()){
        if(array1[p1]<array2[p2]){
            res[index++] = array1[p1++];
        }
        else{
            res[index++] = array2[p2++];
        }
    }
    while (p1<array1.size()){
        res[index++] = array1[p1++];
    }
    while (p2<array2.size()){
        res[index++] = array2[p2++];
    }
    return res;
}
vector<int> mergeSort(vector<int> &array,int start,int end){
    if (start==end){
        vector<int> res;
        res.push_back(array[end]);
        return res;
    }

    int mid = (start+end)/2;
    vector<int> left  = mergeSort(array,start,mid);
    vector<int> right = mergeSort(array,mid+1,end);
    return merge(left,right);
}
vector<int> toVector(int * array,int &size){
    vector<int> res;
    for (int i = 0; i < size; i++){
        res.push_back(array[i]);
    }
    return res;
}
int main(){    
    vector<int> array = {10,9,8,7,6,5,4,3,2,1};
    displayArray(array);

    array = mergeSort(array,0,array.size()-1);
    displayArray(array);

    return 0;
}

File: SEM2/ALGO/ASSIGNMENT 2/mergesort.py
import random
class MergeSort:
    def __init__(self,size) -> None:
        self.arr = self.generateRandomArrayOfSize(size)
        self.count = 0
    def generateRandomArrayOfSize(self,x):
        res = []
        for i in range(x):
            res.append(random.randint(1,100000))
        return res
    def getComparisonCount(self):
        return self.count
    def merge(self, array1, size1, array2, size2):
        res = [None]*(size1+size2)
        p1 = 0
        p2 = 0
        index = 0

        while p1 < size1 and p2 < size2:
            if array1[p1] < array2[p2]:
                res[index] = array1[p1]
                p1 += 1
            else:
                res[index] = array2[p2]
                p2 += 1
            index += 1
            self.count+=1

        while p1 < size1:
            res[index] = array1[p1]
            p1 += 1
            index += 1
            self.count+=1

        while p2 < size2:
            res[index] = array2[p2]
            p2 += 1
            index += 1
            self.count+=1

        return res

    def mergeSort(self, array, start, end):
        if start == end:
            return [array[start]]
        mid = (start + end) // 2
        left = self.mergeSort(array, start, mid)
        right = self.mergeSort(array, mid + 1, end)

        return self.merge(left, len(left), right, len(right))
    def sort(self):
        self.mergeSort(self.arr,0,len(self.arr)-1)
